File: src-tauri/src/wake_word.rs
use std::process::Command;

/// Transcribe a short audio clip using OpenAI's Whisper API.
/// Audio is passed as base64-encoded data (mp4/webm from MediaRecorder).
#[tauri::command]
pub async fn transcribe_audio(audio_base64: String, extension: String) -> Result<String, String> {
    let config = crate::commands::read_config_internal()?;
    let api_key = config
        .api_key
        .ok_or("No API key in ~/.books-reader.json")?;

    let audio_bytes = base64::Engine::decode(
        &base64::engine::general_purpose::STANDARD,
        &audio_base64,
    )
    .map_err(|e| format!("base64 decode: {e}"))?;

    let ext = if extension.is_empty() { "mp4" } else { &extension };
    let tmp_path = format!("/tmp/samuel-wake-audio.{ext}");
    std::fs::write(&tmp_path, &audio_bytes)
        .map_err(|e| format!("write temp audio: {e}"))?;

    // Save a debug copy so we can listen to what was captured
    let debug_path = format!("/tmp/samuel-wake-debug.{ext}");
    let _ = std::fs::copy(&tmp_path, &debug_path);

    eprintln!(
        "[wake] transcribing {:.1}KB audio ({ext}) — debug copy at {debug_path}",
        audio_bytes.len() as f64 / 1024.0
    );

    // Prompt priming: Whisper uses this to expect specific vocabulary,
    // dramatically improving detection of "Hey Samuel" over hallucinations.
    let output = Command::new("curl")
        .args([
            "-s",
            "--max-time",
            "10",
            "-X",
            "POST",
            "https://api.openai.com/v1/audio/transcriptions",
            "-H",
            &format!("Authorization: Bearer {api_key}"),
            "-F",
            &format!("file=@{tmp_path}"),
            "-F",
            "model=whisper-1",
            "-F",
            "language=en",
            "-F",
            "prompt=Hey Samuel",
        ])
        .output()
        .map_err(|e| format!("curl error: {e}"))?;

    let _ = std::fs::remove_file(&tmp_path);

    if !output.status.success() {
        let stderr = String::from_utf8_lossy(&output.stderr);
        return Err(format!("Whisper API error: {stderr}"));
    }

    let body: serde_json::Value =
        serde_json::from_slice(&output.stdout).map_err(|e| format!("parse response: {e}"))?;

    if let Some(err) = body.get("error") {
        return Err(format!("Whisper API: {}", err["message"].as_str().unwrap_or("unknown")));
    }

    let text = body["text"].as_str().unwrap_or("").to_string();
    eprintln!("[wake] whisper: \"{text}\"");
    Ok(text)
}

File: electron/handlers/wake-word.ts
import { writeFileSync, unlinkSync, copyFileSync } from "node:fs";
import { openaiFetch } from "./openai-client.js";

export async function transcribe_audio(args: {
	audio_base64: string;
	extension: string;
}): Promise<string> {
	if (!args.audio_base64) return "";

	const audioBytes = Buffer.from(args.audio_base64, "base64");

	const ext = args.extension || "mp4";
	const tmpPath = `/tmp/samuel-wake-audio.${ext}`;
	writeFileSync(tmpPath, audioBytes);

	const debugPath = `/tmp/samuel-wake-debug.${ext}`;
	try {
		copyFileSync(tmpPath, debugPath);
	} catch {
		// ignore
	}

	console.error(
		`[wake] transcribing ${(audioBytes.length / 1024).toFixed(1)}KB audio (${ext}) — debug copy at ${debugPath}`,
	);

	const form = new FormData();
	form.append(
		"file",
		new Blob([audioBytes], { type: `audio/${ext}` }),
		`audio.${ext}`,
	);
	form.append("model", "gpt-4o-mini-transcribe");
	form.append("language", "en");
	form.append(
		"prompt",
		"Transcribe any English speech. If the audio is silence, noise, music, or non-English speech, return an empty string.",
	);

	const resp = await openaiFetch("/v1/audio/transcriptions", {
		method: "POST",
		body: form,
		signal: AbortSignal.timeout(10_000),
	});

	try {
		unlinkSync(tmpPath);
	} catch {
		// ignore
	}

	if (!resp.ok) {
		const errText = await resp.text();
		throw new Error(`Whisper API error: ${errText}`);
	}

	const body = await resp.json();

	if (body.error) {
		throw new Error(
			`Whisper API: ${body.error.message ?? "unknown"}`,
		);
	}

	const text: string = body.text ?? "";
	console.error(`[wake] whisper: "${text}"`);
	return text;
}
